identity-did/src/document/core_document.rs

```diff
@@ -673,32 +711,32 @@ mod tests {
 
     let fragment = "#detach-test";
     let method = method(document.id(), fragment);
-    document.insert_method(method, MethodScope::VerificationMethod);
+    document.insert_method(method, MethodScope::VerificationMethod).unwrap();
 
     assert!(document
       .attach_method_relationship(
         document.id().to_url().join(fragment).unwrap(),
         MethodRelationship::AssertionMethod,
       )
-      .unwrap());
+      .is_ok());
 
     assert!(document
       .detach_method_relationship(
         document.id().to_url().join(fragment).unwrap(),
         MethodRelationship::AssertionMethod,
       )
-      .unwrap());
+      .is_ok());
 
     // len is 1; the relationship was removed.
     assert_eq!(document.verification_relationships().count(), 3);
 
-    // Removing it a second time returns Ok(false).
-    assert!(!document
+    // Removing it a second time is not an error (idempotent).
+    assert!(document
       .detach_method_relationship(
         document.id().to_url().join(fragment).unwrap(),
         MethodRelationship::AssertionMethod,
       )
-      .unwrap());
+      .is_ok());
```

Comment:
It's consistent within the respective operations, i.e. insertion and removal of methods on the one hand, and modifying relationships on the other. The problem with making the insertion or removal idempotent is that the check for existence in the `OrderedSet` is based on the key. Adding a method with the same fragment, but different `MethodData`s, would either overwrite the existing method or drop the new one. We could go with the former, just how `HashMap::insert` works.

Removal could be made idempotent, of course. Typically, I think you would call remove on something you expect to exist, and if it doesn't, that may be a sign of something going wrong. One can always ignore the resulting `MethodNotFound` error to get the idempotent behavior.

For the low-level API we could easily go with the overwriting (insert) and idempotent (removal) variant, but it's not clear to me what behavior we should strive for in the account.

Reply:
The standard for `OrderedSet` defines inserting/appending a duplicate key as a no-op: https://infra.spec.whatwg.org/#ordered-set
We could change that at a higher level (like how we're returning an error with `insert_method`) but I feel like overwriting a method might be unexpected. We could have a dedicated `replace_method` function I think, which would update the key but retain the relationships etc. (future PR).

In general we want hand-holding errors for the `Account` but because `attach/detach`-ing relationships is low-risk (it prevents removing embedded methods) relative to `insert_method/remove_method`, having them idempotent is probably fine for both?

Reply:
I agree, going with the idempotent behavior for relationships for both APIs seems fine.

It's a bit unclear to me, where that leaves insertion and removal in `CoreDocument`. If we want to cut down on error types as much as possible, we can choose to go with the no-op for insertion and remove `MethodAlreadyExists` from the `identity-did` crate. It'd still be needed in `account`, though.

Reply:
I say we leave method insertion/removal in `CoreDocument` as non-idempotent, if only to prevent silent bugs from trying to insert methods with the same fragment. That makes it easier for the `Account` too.